lavc/qsvdec: Allow decoders to export crop information

feiwan1 opened the pull request:
Correct frame resolution when it change inside a sequence which has max width/height in its syntax. In this case, it will not trigger reinit process, and use the 1st frame's resolution for all frames.

feiwan1 committed:
lavc/qsvdec: Allow decoders to export crop information

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

xhaihao commented on libavcodec/qsvdec.c:
```diff
@@ -851,6 +851,10 @@ static int qsv_decode(AVCodecContext *avctx, QSVContext *q,
+        if (frame->width != outsurf->Info.CropW || frame->height != outsurf->Info.CropH) {
+            avctx->width  = frame->width  = outsurf->Info.CropW;
+            avctx->height = frame->height = outsurf->Info.CropH;
+        }
```
widthxheight is the coded dimensions (in pixel) which is different from the cropping rectangle, see https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/frame.h#L437-L446

feiwan1 replied:
OK, export crop props seems a more generic way, which will also take care of real crop case. I will try to modify it.

xhaihao commented on libavutil/hwcontext_qsv.c:
```diff
@@ -1317,8 +1317,6 @@ static int qsv_get_buffer(AVHWFramesContext *ctx, AVFrame *frame)
     frame->format  = AV_PIX_FMT_QSV;
-    frame->width   = ctx->width;
-    frame->height  = ctx->height;
 
```
qsv_get_buffer is used in qsv based filters too, does this change impact these filters ?

feiwan1 committed:
lavc/qsvdec: Use coded_w/h for frame resolution when use system memory

Fix output mismatch when decode clip with crop(conf_win_*offset in
syntax) info by using system memory:

$ ffmpeg -c:v hevc_qsv -i conf_win_offet.bit -y out.yuv

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

xhaihao reviewed (approved):
LGTM

feiwan1 commented:
https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=11837
